src/main/java/eu/darkbot/captcha/CaptchaSolver.java

```diff
@@ -20,106 +22,133 @@
 import java.util.Collections;
 import java.util.Map;
 import java.util.Set;
-import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
 
 public class CaptchaSolver implements CaptchaAPI {
 
     private static final String url = "https://www.darkorbit.com/";
-    private static String key;
-    private static boolean firstTime;
 
-    private static void initFX() {
-        JFXPanel jfxPanel = new JFXPanel();
-        jfxPanel.setPreferredSize(new Dimension(460, 535));
-        Platform.runLater(() -> {
-            WebView webView = new WebView();
-            jfxPanel.setScene(new Scene(webView));
+    @Override
+    public Map<String, String> solveCaptcha(URL url, String frontPage) {
+        if (!url.toString().contains("lp") || !frontPage.contains("recaptcha_image")) return Collections.emptyMap();
+
+        System.out.println("Captcha detected, opening solver...");
+
+        CompletableFuture<String> key = new CompletableFuture<>();
+        SwingUtilities.invokeLater(() -> new SolverJFXPanel(key));
+
+        try {
+            String response = key.get();
+            if (response != null) return Collections.singletonMap("g-recaptcha-response", response);
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+        }
+
+        System.out.println("No solution to the captcha was provided, timed out.");
+        return Collections.emptyMap();
+    }
+
+    private static class SolverJFXPanel extends JFXPanel {
+        private final CompletableFuture<String> key;
+        private final WebView webView;
+
+        private Timeline timeline;
+        private JDialog dialog;
+        private int maxFrames = 120;
+
+        public SolverJFXPanel(CompletableFuture<String> key) {
+            this.key = key;
+            this.webView = new WebView();
+            setPreferredSize(new Dimension(460, 535));
+
+            setScene(new Scene(webView));
```

Comment:
this line need to be inside `Platform.runLater` thread

Reply:
I moved this to here because according to the JFXPanel documentation, setScene can be called from either the JavaFX thread or the Swing Event Thread <https://docs.oracle.com/javase/8/javafx/api/javafx/embed/swing/JFXPanel.html#setScene-javafx.scene.Scene->
> Attaches a Scene object to display in this JFXPanel. This method can be called either on the event dispatch thread or the JavaFX application thread.

Also the Scene documentation (for calling `new Scene(webView)`) states:
<https://openjfx.io/javadoc/18/javafx.graphics/javafx/scene/Scene.html>
> A Scene may be created and modified on any thread until it is attached to a [Window](https://openjfx.io/javadoc/18/javafx.graphics/javafx/stage/Window.html) that is [showing](https://openjfx.io/javadoc/18/javafx.graphics/javafx/stage/Window.html#isShowing()). After that, it must be modified only on the JavaFX Application Thread. Note that Scene is not thread-safe; modifying a Scene on multiple threads at the same time will lead to unpredictable results and must be avoided.